### ISSUE TITLE
File::Spec->case_tolerant() should return true on Cygwin

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8555]
Migrated from [rt.perl.org#40103](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40103) (status was 'resolved')

Searchable as RT40103$

**Comment by p5pRT** [comment]
### From g.paulissen@chello.nl

This is a bug report for perl from g\.paulissen@&#8203;chello\.nl\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.4\.




---
In think the File&#8203;::Spec&#8203;::Cygwin case\_tolerant\(\) should return true \(1\) instead of false \(0\)\.

$ perl  \-e "use File&#8203;::Spec; print File&#8203;::Spec\->case\_tolerant\(\);"
0

Why?

Have a look at the following commands&#8203;:

$ which sqlplus
/cygdrive/c/Oracle/product/10\.1\.0/Db\_\_4/bin/sqlplus

$ ln \-s /cygdrive/c/Oracle/product/10\.1\.0/Db\_\_4/bin/sqlplus SQLPLUS\.exe

$ PATH=\.&#8203;:$PATH

$ which sqlplus
\./sqlplus

So although the executable is named SQLPLUS\.exe\, which still can find it\.



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.4&#8203;:

Configured by ActiveState at Tue Jun  1 11&#8203;:52&#8203;:09 2004\.

Summary of my perl5 \(revision 5 version 8 subversion 4\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags ='\-nologo \-Gf \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32 \-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT  \-DNO\_HASH\_SEED \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO \-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Progra~1\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth=C&#8203;:\\PROGRA~1\\MICROS~3\\VC98\\lib
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl58\.lib
&nbsp;    gnulibc\_version='undef'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Progra~1\\Perl\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;    ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
&nbsp;    22751 Update to Test\.pm 1\.25
&nbsp;    21540 Fix backward\-compatibility issues in if\.pm



---
@&#8203;INC for perl v5\.8\.4&#8203;:
&nbsp;    h&#8203;:\\work\\isr\\scripts
&nbsp;    h&#8203;:\\work\\isr\\build\\ISR\\scripts
&nbsp;    C&#8203;:/Program Files/Perl/lib
&nbsp;    C&#8203;:/Program Files/Perl/site/lib
&nbsp;    \.



---
Environment for perl v5\.8\.4&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=h&#8203;:\\work\\isr\\scripts;h&#8203;:\\work\\isr\\build\\ISR\\scripts;h&#8203;:\\work\\tw\\common\\src\\perl;C&#8203;:\\Oracle\\product\\10\.1\.0\\Db\_\_4\\bin;C&#8203;:\\Oracle\\product\\10\.1\.0\\Db\_\_4\\jre\\1\.4\.2\\bin\\client;C&#8203;:\\Oracle\\product\\10\.1\.0\\Db\_\_4\\jre\\1\.4\.2\\bin;C&#8203;:\\oracle\\product\\10\.2\.0\\client\_1\\bin;k&#8203;:\\oap\\isr\\dev\\ant\\bin;c&#8203;:\\usr\\local\\wbin;C&#8203;:\\oraclexe\\app\\oracle\\product\\10\.2\.0\\server\\bin;C&#8203;:\\Progra~1\\Perl\\bin;C&#8203;:\\DS10G\\jre\\1\.1\.8\\bin;C&#8203;:\\DS10G\\jlib;C&#8203;:\\DS10G\\jdk\\jre\\bin\\client;C&#8203;:\\DS10G\\jdk\\jre\\bin;C&#8203;:\\DS10G\\jdk\\jre\\bin\\classic;C&#8203;:\\DS10G\\jre\\1\.4\.1\\bin;C&#8203;:\\DS10G\\bin;C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;C&#8203;:\\ORANT\\bin;C&#8203;:\\Program Files\\Oracle\\jre\\1\.1\.7\\bin;C&#8203;:\\ORANT\\jdk\\bin;C&#8203;:\\ORANT\\REPADM61\\jre1\_4\\bin;C&#8203;:\\ORANT\\REPADM61\\jre1\_4\\bin\\client;C&#8203;:\\DES\_817\\bin;C&#8203;:\\Program Files\\Rumba\\Windows;C&#8203;:\\Program Files\\ora817\\bin;C&#8203;:\\Program Files\\Java\\j2sdk1\.4\.2\_10\\bin;C&#8203;:\\Progra~1\\ora817\\bin;C&#8203;:\\Program Files\\COMMON~1\\Crystal Decisions\\2\.0\\bin;C&#8203;:\\Program Files\\Hummingbird\\Connectivity\\10\.00\\Accessories;h&#8203;:\\work\\pldoc;c&#8203;:\\progra~1\\putty
&nbsp;    PERLLIB=h&#8203;:\\work\\isr\\scripts;h&#8203;:\\work\\isr\\build\\ISR\\scripts
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @schwern

g\.paulissen@&#8203;chello\.nl \(via RT\) wrote&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> In think the File&#8203;::Spec&#8203;::Cygwin case\_tolerant\(\) should return true \(1\) instead of false \(0\)\.
> 
> $ perl  \-e "use File&#8203;::Spec; print File&#8203;::Spec\->case\_tolerant\(\);"
> 0
> 
> Why?
> 
> Have a look at the following commands&#8203;:
> 
> $ which sqlplus
> /cygdrive/c/Oracle/product/10\.1\.0/Db\_\_4/bin/sqlplus
> 
> $ ln \-s /cygdrive/c/Oracle/product/10\.1\.0/Db\_\_4/bin/sqlplus SQLPLUS\.exe
> 
> $ PATH=\.&#8203;:$PATH
> 
> $ which sqlplus
> \./sqlplus
> 
> So although the executable is named SQLPLUS\.exe\, which still can find it\.
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Is that magic in which or is that magic in the filesystem?  I know Cygwin's which is already pretty magical with regard to \.exe\.

A simple way to test this is&#8203;:

open FILE\, ">this is a test file" or die $\!;
print FILE "wibble\\n";
close FILE;

print open\(FILE\, "This is a Test FILE"\) ? "Tolerant" : "Intolerant";

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rurban

Attached patch fixes \#40103 for blead\.
\-\- 
Reini Urban

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>pl-#40103-File-Spec-case_tolerant.patch</summary>

```diff
--- perl-5.9.5/lib/File/Spec/Cygwin.pm.orig	2006-11-23 03:13:19.000000000 +0000
+++ perl-5.9.5/lib/File/Spec/Cygwin.pm	2007-06-20 17:25:13.625000000 +0000
@@ -4,7 +4,7 @@
 use vars qw(@ISA $VERSION);
 require File::Spec::Unix;
 
-$VERSION = '1.1';
+$VERSION = '1.2';
 
 @ISA = qw(File::Spec::Unix);
 
@@ -97,6 +97,15 @@
     $tmpdir = $_[0]->_tmpdir( $ENV{TMPDIR}, "/tmp", 'C:/temp' );
 }
 
+=item case_tolerant
+
+Override Unix. Cygwin is always case-tolerant, indicating that it is not
+significant when comparing file specifications.
+
+=cut
+
+sub case_tolerant () { 1 }
+
 =back
 
 =head1 COPYRIGHT

```
</details>

**Comment by p5pRT** [comment]
### From @rurban

On Wed Jun 20 10&#8203;:27&#8203;:50 2007\, rurban wrote&#8203;:
> Attached patch fixes \#40103 for blead\.

Oops\, forgot the test\. 
The attached patch fixes \#40103 for blead and corerects the test\.

\-\- 
Reini Urban

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>pl-#40103-File-Spec-case_tolerant.patch</summary>

```diff
--- perl-5.9.5/lib/File/Spec/Cygwin.pm.orig	2006-11-23 03:13:19.000000000 +0000
+++ perl-5.9.5/lib/File/Spec/Cygwin.pm	2007-06-20 17:25:13.625000000 +0000
@@ -4,7 +4,7 @@
 use vars qw(@ISA $VERSION);
 require File::Spec::Unix;
 
-$VERSION = '1.1';
+$VERSION = '1.2';
 
 @ISA = qw(File::Spec::Unix);
 
@@ -97,6 +97,15 @@
     $tmpdir = $_[0]->_tmpdir( $ENV{TMPDIR}, "/tmp", 'C:/temp' );
 }
 
+=item case_tolerant
+
+Override Unix. Cygwin is always case-tolerant, indicating that it is not
+significant when comparing file specifications.
+
+=cut
+
+sub case_tolerant () { 1 }
+
 =back
 
 =head1 COPYRIGHT
--- perl-5.9.5/lib/File/Spec/t/Spec.t.orig	2006-10-12 15:10:26.000000000 +0000
+++ perl-5.9.5/lib/File/Spec/t/Spec.t	2007-06-20 18:48:37.406250000 +0000
@@ -618,7 +618,7 @@
 #[ "Epoc->canonpath('/a/.')",                                  '/a'        ],
 #[ "Epoc->canonpath('/.')",                                    '/'         ],
 
-[ "Cygwin->case_tolerant()",         '0'  ],
+[ "Cygwin->case_tolerant()",         '1'  ],
 [ "Cygwin->catdir('/','d2/d3')",     '/d2/d3'  ],
 
 ) ;

```
</details>

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

On 20/06/07\, Reini Urban via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Wed Jun 20 10&#8203;:27&#8203;:50 2007\, rurban wrote&#8203;:
> > Attached patch fixes \#40103 for blead\.
>
> Oops\, forgot the test\.
> The attached patch fixes \#40103 for blead and corerects the test\.

Thanks\, applied\, bumping the version to 1\.1\_01\.

**Comment by p5pRT** [comment]
### From @wb8tyw

Reini Urban via RT wrote&#8203;:
> Attached patch fixes \#40103 for blead\.

Isn't the case tolerant setting something that could be on the file 
system that Cygwin must honor?  Or is Cygwin forcing the setting to be 
insensitive?

Windows has the option of setting the file system to be case sensitive\. 
&nbsp;  I do not know at the present time if it is on a per process and 
children setting or if it is a file system or even directory property\.

On VMS it is a per\-process setting\, but only affects the ODS\-5 file system\.

\-John
wb8tyw@&#8203;qsl\.net
Personal Opinion Only

**Comment by p5pRT** [comment]
### From @rurban

On Thu Jun 21 01&#8203;:39&#8203;:59 2007\, rafael wrote&#8203;:
> On 20/06/07\, Reini Urban via RT \<perlbug\-followup \<\!\-\- x \-\-> at
perl\.org> wrote&#8203;:
> > On Wed Jun 20 10&#8203;:27&#8203;:50 2007\, rurban wrote&#8203;:
> > > Attached patch fixes \#40103 for blead\.
> >
> > Oops\, forgot the test\.
> > The attached patch fixes \#40103 for blead and corerects the test\.
> 
> Thanks\, applied\, bumping the version to 1\.1\_01\.

And this patch is for maint and bumps the version to 1\.1\_02

This also adds "Handle network path names beginning with double slash"
from blead\.
And I added another check for $ENV\{'TMP'\}\, and $ENV\{'TEMP'\} in tempdir
See http&#8203;://cygwin\.com/ml/cygwin/2007\-07/msg00119\.html
 
\-\- 
Reini Urban

**Comment by p5pRT** [comment]
### From @rurban

<details><summary>pl-40103-File-Spec-case_tolerant_2.patch</summary>

```diff
difforig perl-5.8.8/lib/File/Spec

2007-07-08  Reini Urban <rurban@x-ray.at>

diff -ub  perl-5.8.8/lib/File/Spec/Cygwin.pm.orig
--- perl-5.8.8/lib/File/Spec/Cygwin.pm.orig	2005-05-06 16:03:10.000000000 +0000
+++ perl-5.8.8/lib/File/Spec/Cygwin.pm	2007-07-08 11:35:13.227000000 +0000
@@ -4,7 +4,7 @@
 use vars qw(@ISA $VERSION);
 require File::Spec::Unix;
 
-$VERSION = '1.1';
+$VERSION = '1.1_02';
 
 @ISA = qw(File::Spec::Unix);
 
@@ -40,7 +40,25 @@
 sub canonpath {
     my($self,$path) = @_;
     $path =~ s|\\|/|g;
-    return $self->SUPER::canonpath($path);
+
+    # Handle network path names beginning with double slash
+    my $node = '';
+    if ( $path =~ s@^(//[^/]+)(?:/|\z)@/@s ) {
+        $node = $1;
+    }
+    return $node . $self->SUPER::canonpath($path);
+}
+
+sub catdir {
+    my $self = shift;
+
+    # Don't create something that looks like a //network/path
+    if ($_[0] and ($_[0] eq '/' or $_[0] eq '\\')) {
+        shift;
+        return $self->SUPER::catdir('', @_);
+    }
+
+    $self->SUPER::catdir(@_);
 }
 
 =pod
@@ -66,6 +84,8 @@
 
     $ENV{TMPDIR}
     /tmp
+    $ENV{'TMP'}
+    $ENV{'TEMP'}
     C:/temp
 
 Since Perl 5.8.0, if running under taint mode, and if the environment
@@ -76,9 +96,18 @@
 my $tmpdir;
 sub tmpdir {
     return $tmpdir if defined $tmpdir;
-    $tmpdir = $_[0]->_tmpdir( $ENV{TMPDIR}, "/tmp", 'C:/temp' );
+    $tmpdir = $_[0]->_tmpdir( $ENV{TMPDIR}, "/tmp", $ENV{'TMP'}, $ENV{'TEMP'}, 'C:/temp' );
 }
 
+=item case_tolerant
+
+Override Unix. Cygwin is always case-tolerant, indicating that it is not
+significant when comparing file specifications.
+
+=cut
+
+sub case_tolerant () { 1 }
+
 =back
 
 =head1 COPYRIGHT
diff -ub  perl-5.8.8/lib/File/Spec/t/Spec.t.orig
--- perl-5.8.8/lib/File/Spec/t/Spec.t.orig	2005-08-27 17:26:19.000000000 +0000
+++ perl-5.8.8/lib/File/Spec/t/Spec.t	2007-07-08 12:12:06.242625000 +0000
@@ -595,7 +595,7 @@
 #[ "Epoc->canonpath('/a/.')",                                  '/a'        ],
 #[ "Epoc->canonpath('/.')",                                    '/'         ],
 
-[ "Cygwin->case_tolerant()",         '0'  ],
+[ "Cygwin->case_tolerant()",         '1'  ],
 
 ) ;
 

```
</details>

**Comment by p5pRT** [comment]
### From @rurban

On Thu Jun 21 19&#8203;:52&#8203;:13 2007\, malmberg wrote&#8203;:
> Reini Urban via RT wrote&#8203;:
> > Attached patch fixes \#40103 for blead\.
> 
> Isn't the case tolerant setting something that could be on the file 
> system that Cygwin must honor?  Or is Cygwin forcing the setting to be 
> insensitive?
> 
> Windows has the option of setting the file system to be case sensitive\. 
>   I do not know at the present time if it is on a per process and 
> children setting or if it is a file system or even directory property\.
> 
> On VMS it is a per\-process setting\, but only affects the ODS\-5 file
system\.

This is true\, but rarely used and impossible to check \(yet\)\.
We really should check for a managed mount\, and for some registry
setting \(?\) where the case\-sensitivity can be overruled\.
The registry API is not in CORE and a mount API also not \(yet\)\.

For the managed mount I have to extend the Cygwin API for something like
Cygwin&#8203;::mount\_flags\(pathname\, \[flag\]\) => @&#8203; or BOOL
flag and/or return array one of 
qw\(text binary managed user system exec notexec cygexec nosuid\)
\-\- 
Reini Urban
